### PR TITLE
control/controlclient: flesh out mapSession to break up gigantic method

### PR DESCRIPTION
**Pull request by bradfitz**
Now mapSession has a bunch more fields and methods, rather than being
just one massive func with a ton of local variables.

So far there are no major new optimizations, though. It should behave
the same as before.

This has been done with an eye towards testability (so tests can set
all the callback funcs as needed, or not, without a huge Direct client
or long-running HTTP requests), but this change doesn't add new tests
yet. That will follow in the changes which flesh out the NetmapUpdater
interface.

Updates #1909

### CODE REVIEW

File: control/controlclient/direct.go [hunk 1]
```diff
@@ -1539,6 +1503,9 @@ func answerC2NPing(logf logger.Logf, c2nHandler http.Handler, c *http.Client, pr
 }
 
 func sleepAsRequested(ctx context.Context, logf logger.Logf, timeoutReset chan<- struct{}, d time.Duration, clock tstime.Clock) error {
+	if timeoutReset == nil {
+		timeoutReset = make(chan struct{})
+	}
```
Comment (maisem) [inline control/controlclient/direct.go:1508]: this is confusing, whats the point of making a channel here that no one can read from and always blocks anyway?

File: control/controlclient/map.go [hunk 2]
```diff
@@ -49,23 +77,121 @@ type mapSession struct {
...
+	lastNetmapSummary      string // from NetworkMap.VeryConcise
 
 	// netMapBuilding is non-nil during a netmapForResponse call,
 	// containing the value to be returned, once fully populated.
 	netMapBuilding *netmap.NetworkMap
 }
 
-func newMapSession(privateNodeKey key.NodePrivate) *mapSession {
+// newMapSession returns a new mapSession.
+//
```
Comment (maisem) [inline control/controlclient/map.go:88]: say that fields like cancel, onConciseNetMapSummary are stubs and should be replaced?

File: control/controlclient/direct.go (hunk 1 above)
Comment (bradfitz) [reply control/controlclient/direct.go:1508]: Oh, I was wrong about nil writable channels in select. Will remove. (They're fine: https://go.dev/play/p/po5IcntgwFU)

Comment (maisem) [reply control/controlclient/direct.go:1508]: yeah, they just block forever

File: control/controlclient/direct.go [hunk 3]
```diff
@@ -806,10 +805,10 @@ func (c *Direct) SendUpdate(ctx context.Context) error {
 	return c.sendMapRequest(ctx, false, nil)
 }
 
-// If we go more than pollTimeout without hearing from the server,
+// If we go more than watchdogTimeout without hearing from the server,
 // end the long poll. We should be receiving a keep alive ping
 // every minute.
-const pollTimeout = 120 * time.Second
+const watchdogTimeout = 120 * time.Second
```
Comment (maisem) [inline control/controlclient/direct.go:811]: oh this timeout lol

File: control/controlclient/direct.go [hunk 4]
```diff
@@ -1074,82 +1078,42 @@ func (c *Direct) sendMapRequest(ctx context.Context, isStreaming bool, nu Netmap
...
-		if nm.SelfNode == nil {
-			c.logf("MapResponse lacked node")
-			return errors.New("MapResponse lacked node")
-		}
-
-		if DevKnob.StripEndpoints() {
-			for _, p := range resp.Peers {
-				p.Endpoints = nil
+		} else {
+			if resp.Node == nil {
```
Comment (maisem) [inline control/controlclient/direct.go:1084]: if the first message you get is a keep alive, followed up by a map response which doesn't have the SelfNode this check won't fire

Comment (bradfitz) [reply control/controlclient/direct.go:1084]: I don't think that's legal. The very first MapResponse must be a full one.

File: control/controlclient/direct.go [hunk 5]
```diff
@@ -1074,82 +1082,42 @@ func (c *Direct) sendMapRequest(ctx context.Context, isStreaming bool, nu Netmap
 		}
 
 		metricMapResponseMap.Add(1)
-		if i > 0 {
+		if mapResIdx > 0 {
 			metricMapResponseMapDelta.Add(1)
```
Comment (maisem) [inline control/controlclient/direct.go:1086]: this isn't necessarily delta until after `gotNonKeepAliveMessage = true`

**Review by maisem — APPROVED** [review]

**Pull request merged**